### PR TITLE
UHF Hessian and gradients

### PR DESCRIPTION
**Pull request by jeffschriber**
## Description
Adds analytic UHF hessians

## Todos
Notable points that this PR has either accomplished or will accomplish.

  - [X] UHF response terms (conventional algorithm)
  - [X] UHF response terms (DF algorithm)
  - [X] UHF JK derivs
  - [X] Add tests 

## Status
- [x] Ready for review
- [x] Ready for merge

### CODE REVIEW

File: psi4/src/psi4/scfgrad/scf_grad.cc [hunk 1]
```diff
@@ -1494,7 +1495,7 @@ SharedMatrix SCFGrad::compute_hessian()
     if (options_.get_str("REFERENCE") == "RHF") {
         hessians_["Response"] = rhf_hessian_response();
     } else {
-        throw PSIEXCEPTION("SCFHessian: Response not implemented for this reference");
+        hessians_["Response"] = uhf_hessian_response();
     }
```
[the remark this answers is not in the document]
Comment (loriab) [reply psi4/src/psi4/scfgrad/scf_grad.cc]: Sorry, I think the above "else {" comment has been lingering on GH for months. I don't understand it myself now.

File: tests/scf-hess2/input.dat [hunk 2]
```diff
@@ -0,0 +1,89 @@
+#! UHF STO-3G (Cartesian) and cc-pVDZ (spherical) water Hessian test, against Psi3 reference values.
+#! This test should match RHF values exactly
+
+psi3_hess = np.array([                                                                                    #TEST
```
Comment (loriab) [inline tests/scf-hess2/input.dat]: be sure to add the new tests to `tests/CMakeLists.txt` so they run.

Comment (jeffschriber) [reply tests/scf-hess2/input.dat]: Ah yes, thanks!

**Comment by loriab** [comment]
ping that this needs a rebase to resolve a conflict.

does anything special need to be added for UHF dipole derivatives?

it'd be nice to get an open-shell in the test_vib_analysis II/III/IV section, at least checking hessian and  DD match for dertype=2/1/0. that can be a follow-up PR, but if you have reference Hessian and other arrays lying around from another code, let's collect them into that file. (it's a convoluted file to handle different programs' std orientations, so if you wade in, feel free to ping me for a map.)

**Comment by loriab** [comment]
Is this both conventional and DF, like the RHF Hessians? I'm adding a note to the paper to consult with you so assertion can be deleted if wrong.

**Comment by loriab** [comment]
nm, I knew I had seen evidence https://github.com/psi4/psi4/pull/842/files#diff-f816279d0193fcb2e124041a6b60eaf5R19

File: tests/scf-hess4/input.dat [hunk 3]
```diff
@@ -0,0 +1,39 @@
+#! CONV SCF 6-31G analytical vs finite-difference tests
+#! Tests UHF hessian code for Ca != Cb
```
Comment (susilehtola) [inline tests/scf-hess4/input.dat]: Copy-paste error? The orbitals are the same

Comment (jeffschriber) [reply tests/scf-hess4/input.dat]: Yes, good catch. Fixed here and in scf-hess5

File: psi4/src/psi4/scfgrad/jk_grad.cc [hunk 4]
```diff
@@ -1533,37 +1553,77 @@ void DFJKGrad::compute_hessian()
...
+
+                // First alpha
+                // (A|mj) = (A|mn) C[n][j]
+                C_DGEMM('N','N',np*(size_t)nso,na,nso,1.0,Amnp[0],nso,Cap[0],na,0.0,Aa_mip[0],na);
+                // (A|ij) = (A|mj) C[m][i]
+                #pragma omp parallel for
+                for (int p = 0; p < np; p++) {
+                    C_DGEMM('T','N',na,na,nso,1.0,Aa_mip[p],na,Cap[0],na,0.0,&Aa_ijp[0][p * (size_t) na * na],na);
+                }
+                // Beta
```
Comment (andysim) [inline psi4/src/psi4/scfgrad/jk_grad.cc:1567]: This chunk of code needs to be moved outside the loops, as [done here](https://github.com/psi4/psi4/pull/1805/files).  It's correct as-is, but veeeeeerrrrrry slow.

File: psi4/src/psi4/scfgrad/jk_grad.cc [hunk 5]
```diff
@@ -1864,111 +1972,117 @@ void DFJKGrad::compute_hessian()
...
-                                nxnz += Cpmn * buffer[50 * stride + delta];
-                                nyny += Cpmn * buffer[51 * stride + delta];
-                                nynz += Cpmn * buffer[52 * stride + delta];
-                                nznz += Cpmn * buffer[53 * stride + delta];
-                                ++delta;
+                
+                    // Loop through alpha/beta terms
+                    std::vector<double**> Blist = {Ba_mnp};
+                    if (!same_ab) Blist.push_back(Bb_mnp);
+                    for (auto& B : Blist){
```
Comment (andysim) [inline psi4/src/psi4/scfgrad/jk_grad.cc:1979]: This is smart way to handle the unknown beta terms

File: psi4/src/psi4/scfgrad/scf_grad.h [hunk 6]
```diff
@@ -71,6 +73,64 @@ class RSCFDeriv : public SCFDeriv {
...
+                    std::shared_ptr<Matrix> C2occ, 
+                    int nso, int n1occ, int n2occ, int nvir,bool alpha);
+
+    void dipole_derivatives(std::shared_ptr<Matrix> C1, 
+                            std::shared_ptr<Matrix> C1occ,
+                            std::shared_ptr<Matrix> C2, 
+                            std::shared_ptr<Matrix> C2occ,
+                            std::shared_ptr<Matrix> Dt,
+                            int nso, int n1occ, int n2occ, int n1vir);
+
```
Comment (andysim) [inline psi4/src/psi4/scfgrad/scf_grad.h:127]: These helpers are a good idea.  After this is in, I'll rework the RHF code to use them.

File: psi4/src/psi4/scfgrad/jk_grad.cc (hunk 4 above)
Comment (jeffschriber) [reply psi4/src/psi4/scfgrad/jk_grad.cc]: Think this is done now. David mentioned that you had made a change which provided some massive speedups. I wasn't 100% sure if I'd done this work before or after that, so I'm glad you settled it!

File: psi4/src/psi4/scfgrad/response.cc [hunk 7]
```diff
@@ -1548,5 +1549,1878 @@ std::shared_ptr<Matrix> RSCFDeriv::hessian_response()
...
+
+    // => Sizing <= //
+
+    int natom = molecule_->natom();
+    int nso   = basisset_->nbf();
+    int naocc  = eps_aocc->dimpi()[0];
+    int navir  = eps_avir->dimpi()[0];
+    int nbocc  = eps_bocc->dimpi()[0];
+    int nbvir  = eps_bvir->dimpi()[0];
+    int nmo   = Ca->colspi()[0];
```
Comment (loriab) [inline psi4/src/psi4/scfgrad/response.cc:1583]: if convenient, run `clang-format` over this file

File: psi4/src/psi4/scfgrad/response.cc [hunk 8]
```diff
@@ -1548,5 +1549,1878 @@ std::shared_ptr<Matrix> RSCFDeriv::hessian_response()
...
+        pdip_grad[A][2] += 2*mu1_z.vector_dot(U1pi);
+    }
+    for (int A = 0; A < 3*natom; A++) {
+        psio_->read(PSIF_HESS,"Upi^A_b",(char*)U2pip[0], static_cast<size_t> (nmo) * n2occ * sizeof(double),next_Upi,&next_Upi);
+        pdip_grad[A][0] += 2*mu2_x.vector_dot(U2pi);
+        pdip_grad[A][1] += 2*mu2_y.vector_dot(U2pi);
+        pdip_grad[A][2] += 2*mu2_z.vector_dot(U2pi);
+    }
+    uhf_wfn_->set_array_variable("SCF DIPOLE GRADIENT", dipole_gradient);
+    uhf_wfn_->set_array_variable("CURRENT DIPOLE GRADIENT", dipole_gradient);
```
Comment (loriab) [inline psi4/src/psi4/scfgrad/response.cc:3423]: Thanks for including these!

**Pull request merged**